### PR TITLE
Add comprehensive input validation with detailed error responses

### PR DESCRIPTION
**Pull request by olaservo**
## Summary
- Added comprehensive input validation middleware for all Todo endpoints
- Implemented field-specific error messages with structured error responses  
- Added validation for title length, HTML tag prevention, and type checking
- Enhanced ID parameter validation for route security
- Added comprehensive test coverage for all validation scenarios

### New Files
- `src/middleware/validation.ts` - Core validation logic with custom error handling
- `src/middleware/validation.test.ts` - Comprehensive test suite (15+ test cases)

### Modified Files
- `src/routes/todos.ts` - Integrated validation middleware, improved error responses
- `README.md` - Added validation rules and error response format documentation
- `package.json` - Added supertest for API testing

### Validation Rules Implemented
- **Title**: Required, 1-100 chars, no HTML tags, string type validation
- **Description**: Optional, max 500 chars, no HTML tags, string type validation  
- **Completed**: Boolean type validation when provided
- **ID Parameters**: Positive integer validation for all routes

## Test Coverage
- Valid input acceptance
- Missing required fields
- Invalid data types
- Length limit enforcement
- HTML tag prevention (XSS protection)
- Multiple simultaneous errors
- Edge cases and boundary conditions

## API Response Changes
**Before**: Simple error strings
```json
{ "error": "Title is required" }
```

**After**: Structured error objects with field details
```json
{
  "error": "Validation failed", 
  "details": [
    { "field": "title", "message": "Title is required" },
    { "field": "description", "message": "Description must be 500 characters or less" }
  ]
}
```

## Security Improvements
- HTML tag prevention to mitigate XSS attacks
- Input sanitization with automatic trimming
- Type validation to prevent injection attacks
- ID format validation to prevent path traversal

## Testing Instructions

1. **Install dependencies**: `npm install`
2. **Run test suite**: `npm test` (should show 15+ passing tests)
3. **Test API manually**:
   ```bash
   npm run dev
   
   # Test valid creation
   curl -X POST localhost:3000/api/todos \
     -H "Content-Type: application/json" \
     -d '{"title":"Valid Todo","description":"Valid description"}'
   
   # Test validation errors
   curl -X POST localhost:3000/api/todos \
     -H "Content-Type: application/json" \
     -d '{"title":"","description":"'$(printf 'a%.0s' {1..501})'"}'
   ```

## Breaking Changes
**Error response format has changed** - API consumers will need to handle the new structured error format

## Future Enhancements
- [ ] Add rate limiting validation
- [ ] Implement custom validation rules configuration
- [ ] Add internationalization for error messages
- [ ] Consider adding request sanitization middleware

### CODE REVIEW

File: src/middleware/validation.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    errors.push({ field: 'title', message: 'Title cannot contain HTML tags' });
+  }
+
+  // Validate description
+  if (description !== undefined && description !== null) {
+    if (typeof description !== 'string') {
+      errors.push({ field: 'description', message: 'Description must be a string', value: description });
+    } else if (description.length > 500) {
+      errors.push({ field: 'description', message: 'Description must be 500 characters or less' });
+    } else if (/<.*>/g.test(description)) {  // BUG: Inconsistent regex - allows <img> but title rejects it
```
Comment (dandeliongold) [inline src/middleware/validation.ts:42]: **Bug: Inconsistent HTML tag validation regex**

The regex for HTML tag validation is inconsistent between title and description:
- Title uses: `/<[^>]*>/g` (rejects any HTML tags)  
- Description uses: `/<.*>/g` (allows self-closing tags like ``)

This inconsistency could lead to unexpected behavior. Both should use the same regex pattern for consistent validation.

File: src/middleware/validation.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+export const validateTodoId = (req: Request, res: Response, next: NextFunction) => {
+  const { id } = req.params;
+
+  if (!id || !/^\d+$/.test(id)) {
+    return res.status(400).json({
+      error: 'Invalid ID format',
+      details: [{ field: 'id', message: 'ID must be a positive integer', value: id }]
+    });
+  }
+
```
Comment (dandeliongold) [inline src/middleware/validation.ts:109]: **Bug: ID validation allows "0" but may be invalid**

The current regex `^\d+$` allows ID "0", but depending on your ID generation strategy, "0" might not be a valid ID. Consider using `^[1-9]\d*$` if you want to enforce positive integers starting from 1.

File: src/routes/todos.ts [hunk 3]
```diff
@@ -9,54 +10,67 @@ router.get('/', (req: Request, res: Response) => {
...
+router.post('/', validateTodoCreate, (req: Request, res: Response) => {
   const { title, description }: CreateTodoRequest = req.body;
   
-  if (!title || title.trim() === '') {
-    return res.status(400).json({ error: 'Title is required' });
-  }
-  
-  const todo = todoStore.create({ title: title.trim(), description });
+  // BUG: No try-catch block - if todoStore.create throws, server will crash
+  const todo = todoStore.create({
```
Comment (dandeliongold) [inline src/routes/todos.ts:31]: **Security Risk: Missing error handling**

This code lacks try-catch error handling. If `todoStore.create` throws an exception (e.g., database error, validation error), the server will crash. Consider wrapping in a try-catch block:

```typescript
try {
  const todo = todoStore.create({ 
    title: title.trim(), 
    description: description?.trim() || undefined 
  });
  res.status(201).json(todo);
} catch (error) {
  res.status(500).json({ 
    error: 'Internal server error',
    details: [{ message: 'Failed to create todo' }]
  });
}
```

File: src/middleware/validation.test.ts [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        .post('/')
+        .send({ title: '<script>alert("xss")</script>' });
+
+      expect(response.status).toBe(400);
+      expect(response.body.details).toContainEqual({
+        field: 'title',
+        message: 'Title cannot contain HTML tags'
+      });
+    });
+
```
Comment (dandeliongold) [inline src/middleware/validation.test.ts:61]: **Missing Test Coverage**

The test comments indicate missing test cases:
1. Description HTML validation test (to verify consistency with title validation)
2. ID "0" edge case test

Consider adding these tests to ensure comprehensive coverage:

```typescript
it('should reject description with HTML tags', async () => {
  const response = await request(app)
    .post('/')
    .send({ title: 'Valid Title', description: '<script>alert("xss")</script>' });

  expect(response.status).toBe(400);
  expect(response.body.details).toContainEqual({
    field: 'description',
    message: 'Description cannot contain HTML tags'
  });
});
```

**Review by dandeliongold — CHANGES_REQUESTED** [review]
## Review Summary

This PR adds comprehensive input validation to the Todo API, which is a significant security and user experience improvement. The implementation includes proper validation middleware, structured error responses, and good test coverage.

### ✅ **Strengths:**

1. **Comprehensive Validation**: Covers all critical validation scenarios including type checking, length limits, and HTML tag prevention
2. **Security Improvements**: Implements XSS protection through HTML tag validation and input sanitization
3. **Structured Error Responses**: New error format is much more informative and developer-friendly
4. **Good Test Coverage**: 15+ test cases covering various validation scenarios
5. **Proper Middleware Architecture**: Clean separation of concerns with dedicated validation middleware
6. **Documentation**: Updated README with clear validation rules and error format examples

### ⚠️ **Issues to Address:**

1. **Critical Bug**: Inconsistent HTML validation regex between title and description fields
2. **Potential Bug**: ID validation allows "0" which may not be valid depending on ID strategy  
3. **Security Risk**: Missing error handling in route handlers could cause server crashes
4. **Missing Tests**: Some edge cases noted in comments but not implemented

### 📋 **Recommendations:**

1. **Fix regex consistency**: Use the same HTML validation pattern for both title and description
2. **Add error handling**: Wrap database operations in try-catch blocks
3. **Complete test coverage**: Add the missing test cases mentioned in comments
4. **Consider ID validation**: Decide if "0" should be a valid ID and adjust accordingly

### 🎯 **Breaking Changes:**
As noted, the error response format change is breaking and will require API consumer updates. This should be clearly communicated.

Overall, this is a solid improvement to the API's robustness and security. The issues identified are fixable and don't detract from the overall quality of the implementation.
